### PR TITLE
feat(updater): add built-in property-to-package fallback mappings

### PR DESCRIPTION
**Pull request by agneszitte**
GitHub Issue (If applicable): _None — surfaced by canary build audit_

## PR Type

- Feature

## What is the current behavior?

When an MSBuild version property (e.g. `<UnoMaterialVersion>7.0.0-dev.28</UnoMaterialVersion>`) is discovered during XML scan, the updater derives a package ID by stripping `Version`, splitting at camel-case boundaries, lowercasing, and joining with dots (`UnoMaterialVersion` → `uno.material`). This works for properties whose derived ID matches a real, actively-maintained package on NuGet.org.

It silently fails when the derived ID points at a legacy/deprecated package whose version trajectory has diverged from what the project actually consumes. Repos affected in production today (verified across the full canary build audit on `Uno Platform`, `uno-private`, `Uno Rider Addin` projects):

| Repo | Property | Derived ID returns | Real package | Real latest dev |
|---|---|---|---|---|
| Uno.Studio | `UnoMaterialVersion` | `uno.material` 5.8.0-dev.2 (legacy UWP) | `Uno.Material.WinUI` | 7.0.0-dev.33 |
| Uno.Studio | `UnoExtensionsVersion` | `uno.extensions` 3.0.0-dev.1957 (legacy umbrella) | `Uno.Extensions.Core` (family lockstep) | 7.3.0-dev.78 |
| Uno.Toolkit | `UnoThemesVersion` | `uno.themes` 5.8.0-dev.2 (legacy UWP) | `Uno.Themes.WinUI` | 7.0.0-dev.33 |

In each case the canary updater queries the legacy ID, finds a version *older* than what's already in `Directory.Build.props`, treats it as a non-upgrade, and silently skips. Studio's canary failed 10+ consecutive nights with `NU1605` once an upstream Uno.Toolkit canary advanced its `Uno.Material.WinUI ≥ 7.0.0-dev.29` floor past Studio's frozen `7.0.0-dev.28`.

## What is the new behavior?

A small built-in `PropertyPackageMappings` table (private; only `TryGetFallbackMapping` is exposed) that the updater consults **after** the standard per-feed lookup, gated by a *suspicion signal*: the fallback only fires when the standard result is `null` or has a version *strictly older* than the local reference (signal that the local file is tracking the modern variant via property indirection and the derived legacy ID is a frozen line).

Resolution algorithm:

```text
result := standard lookup
if result is null OR reference.Version > result.Version:
  if WellKnownMappings contains ConvertPackageIdToPropertyFragment(reference.Id):
    fallback := lookup with the mapped canonical package ID
    if fallback is not null AND (result is null OR fallback.Version > result.Version):
      result := fallback
return result
```

Initial well-known table:

| Property fragment | Falls back to | Reason |
|---|---|---|
| `UnoMaterial` | `Uno.Material.WinUI` | Legacy `uno.material` (UWP era) at v5.x is stale; modern consumers reference the WinUI 3 variant. |
| `UnoThemes` | `Uno.Themes.WinUI` | Legacy `uno.themes` (UWP era) at v5.x is stale; modern consumers reference the WinUI 3 variant (and `Uno.Simple.WinUI` ships in lockstep). |
| `UnoExtensions` | `Uno.Extensions.Core` | Legacy umbrella `uno.extensions` at v2.x is stale; modern consumers reference the `Uno.Extensions.*` family in lockstep. |

Users who need explicit, non-fallback property→package mappings should continue to use the existing `--updateProperties=` CLI argument (or the `useUpdateProperties` + `updatePropertiesFile` ADO task inputs). The well-known table is purely a built-in safety net for the common cases that every canary repo would otherwise need to configure individually.

**Suspicion gate** prevents a regression risk for repos with direct `<PackageVersion Include="Uno.Material" Version="5.0.13" />` references (e.g. uno.toolkit.ui pattern) where the legacy line is still actively consumed at 5.x. Without the gate the fallback would have rewritten such references to `Uno.Material.WinUI`'s 7.x version, which doesn't exist for the original package ID and would break restore. With the gate, the standard lookup wins for those cases (5.0.13 → 5.8.0-dev.2 on the legacy line) and the fallback stays out of the way. Verified end-to-end against a mixed-fixture (Studio-style property + Toolkit-style direct ref in one tree) and covered by three boundary tests: `GivenLocalBehindStandard_FallbackDoesNotFire`, `GivenLocalEqualsStandard_FallbackDoesNotFire`, `GivenBasePackageNewerThanFallback_FallbackIsNotPreferred`.

See [`specs/property-package-fallback-mappings/spec.md`](https://github.com/unoplatform/nuget.updater/blob/main/specs/property-package-fallback-mappings/spec.md) for the full design, audit methodology, and rollout plan.

## PR Checklist

- [x] Tested code with current supported SDKs (net8.0 + net9.0 — both build and test green)
- [x] Docs have been added/updated — `doc/nuget-updater.md` gets a new "Property → package resolution" section (including the canonical convention comment quoted from real repos); full design lives in `specs/property-package-fallback-mappings/spec.md`. `CHANGELOG.md` updated under `[Unreleased] > ### Added`; new `specs/README.md` documents the convention for future spec contributors to this repo.
- [x] Unit Tests and/or UI Tests for the changes have been added — 16 new tests (6 unit + 10 end-to-end including suspicion-gate boundary tests, `VersionOverrides` short-circuit, and both-base-and-fallback-absent edge case); `Passed!  - Failed: 0, Passed: 64, Skipped: 1, Total: 65`
- [x] Contains **NO** API breaking changes — purely additive (new file + new safety-net logic gated by suspicion check; well-known table is private). Behavior change: properties previously silently-skipped will now auto-bump (this *is* the fix). Existing `--updateProperties=` and `--versionOverrides=` behavior unchanged (verified by `GivenForcedVersionOverride_FallbackIsNotConsulted` test).
- [x] Updated CHANGELOG
- [ ] Associated with an issue — surfaced by canary audit; not tracked in a separate issue

### Rollout (after merge + release)

1. Publish new `Uno.NuGet.Updater.Tool` version to NuGet.org (e.g. `1.2.11`).
2. Bump `nugetUpdaterVersion:` one-liner in each canary YAML across affected repos: uno.studio, Uno.Gallery, Uno.Themes, Uno.Playground, uno.extensions, uno.toolkit.ui, uno.chefs, uno.rider, uno.csharpmarkup, uno.templates.
3. Watch the first canary build in each repo for `Applied property fallback mapping:` log lines and reduced `## Skipped` count.

### Out of scope (flagged in spec for follow-up)

- **uno.csharpmarkup** `Uno.UI` / `Uno.WinUI.Skia.Gtk` skip — *different bug class* (both packages actively maintained on parallel major lines on NuGet.org; not a deprecated-base situation). Needs a content/configuration change in the csharpmarkup repo, not this tool.
- **uno.extensions** canary YAML has a `projectProperties:` line that doesn't exist as a task input — silently no-op since 2025-08. Can be removed after this lands (the well-known table covers their `MauiUnoVersion` / `NonMauiUnoVersion` intent indirectly via `UnoExtensions` → `Uno.Extensions.Core`, and if they need anything more specific the existing `useUpdateProperties:true` + `updatePropertiesFile:` is the right path).
- **`unoplatform/uno/doc/articles/migrating-to-uno-5.md`** uses `$UnoWinUIVersion$` as a `dotnet new` template variable in a csproj snippet — separate repo, separate fix (will be filed independently).

Internal Issue (If applicable): n/a

### CODE REVIEW

File: src/NvGet/Tools/Updater/Entities/PropertyPackageMappings.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		/// Key: property name with the trailing "Version" stripped (e.g. "UnoMaterial" for "UnoMaterialVersion").
+		/// Value: tuple of (canonical package ID to look up instead, human-readable reason for logs).
+		/// </summary>
+		public static readonly IReadOnlyDictionary<string, (string FallbackPackageId, string Reason)> WellKnownMappings =
+			new Dictionary<string, (string, string)>
+			{
+				// Legacy `uno.material` (UWP-era) is at v5.x and stale.
+				// Modern consumers reference `Uno.Material.WinUI` (currently 7.x).
+				["UnoMaterial"] = ("Uno.Material.WinUI", "Base package Uno.Material is outdated; resolving against Uno.Material.WinUI."),
+
```
[the remark this answers is not in the document]
Comment (agneszitte) [reply src/NvGet/Tools/Updater/Entities/PropertyPackageMappings.cs]: Fair catch. Valid on both counts — the  interface contract doesn't prevent a downcast-and-mutate (e.g. `((IDictionary<string, ...>)WellKnownMappings).Add(...)`), and the only consumer needs the `TryGetFallbackMapping` method anyway. Made the field `private` so the table is genuinely immutable from outside; `TryGetFallbackMapping` is now the only supported access path. Tightens the public API surface to exactly what's required.

File: src/NvGet/Tools/Updater/Extensions/UpdaterParametersExtension.cs [hunk 2]
```diff
@@ -124,6 +170,24 @@ PackageReference reference
 				.FirstOrDefault();
 		}
 
+		/// <summary>
+		/// Inverse of the property-name → package-id heuristic used during XML discovery:
+		/// converts "uno.extensions" back to "UnoExtensions" so the well-known fallback map
+		/// (keyed by property fragment) can be looked up.
```
[the remark this answers is not in the document]
Comment (agneszitte) [reply src/NvGet/Tools/Updater/Extensions/UpdaterParametersExtension.cs]: Valid — the wording oversold what the helper does. The forward heuristic in `XmlDocumentExtensions.GetPackageReferences` lowercases everything, so acronym information is lost on the forward path. Both `UnoWinUIVersion` → `uno.win.ui` and `UnoWinuiVersion` → `uno.winui` are valid forward conversions but this helper can only produce a single output per input (`UnoWinUi` and `UnoWinui` respectively).

Fixed in commit [`1ca609d`](https://github.com/unoplatform/nuget.updater/pull/40/commits/1ca609d): reworded the XML doc to call it what it is — best-effort title-casing for the fallback-table lookup, not a true inverse. Added a worked acronym example in the `<remarks>` and noted that all current well-known entries (`UnoMaterial`, `UnoThemes`, `UnoExtensions`) avoid the ambiguity.

File: doc/nuget-updater.md [hunk 3]
```diff
@@ -127,3 +127,30 @@ The nuget updater supports updating:
...
+
+When a version is tracked via an MSBuild property (e.g. `<UnoMaterialVersion>7.0.0-dev.28</UnoMaterialVersion>`), the updater resolves the corresponding NuGet package ID in two complementary ways:
+
+1. **Explicit mapping (recommended for repo-specific cases).** Pass a JSON file via `--updateProperties=` listing `(PropertyName, PackageId)` pairs. See the `properties.json` example above. This is the authoritative path — the updater will treat the property's value as the version of the named package and update both in lockstep.
+
+2. **Name-derivation heuristic.** When no explicit mapping is provided, the updater derives a package ID from the property name by stripping the trailing `Version`, splitting at lowercase→uppercase boundaries, lowercasing, and joining with dots:
+
+   | Property | Derived package ID |
+   |---|---|
+   | `UnoWinuiVersion` | `uno.winui` |
```
[the remark this answers is not in the document]
Comment (agneszitte) [reply doc/nuget-updater.md:141]: Valid — and worth flagging that the two casings produce **different** derived IDs, not just a stylistic preference: `UnoWinUIVersion` → `uno.win.ui` while `UnoWinuiVersion` → `uno.winui`. Real-world Uno repos use the lowercase-acronym form (matches what the modern `Uno.WinUI` family publishes on NuGet.org, e.g. Studio's `Directory.Build.props` uses `UnoWinuiVersion`).

Fixed in commit [`1ca609d`](https://github.com/unoplatform/nuget.updater/pull/40/commits/1ca609d): standardized both occurrences (line 125 and the new table on line 141) on `UnoWinuiVersion`, and added a *Property name casing matters* note under the derivation table explaining that the heuristic splits at every lowercase→uppercase transition, with a concrete `WinUI` vs `Winui` example so future readers don't trip on it.

File: src/NvGet.Tests/Tools/Updater/PropertyPackageMappingsTests.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NvGet.Tools.Updater.Entities;
+
+namespace NvGet.Tests.Tools.Updater
+{
+	[TestClass]
+	public class PropertyPackageMappingsTests
+	{
+		[TestMethod]
```
Comment (Xiaoy312) [inline src/NvGet.Tests/Tools/Updater/PropertyPackageMappingsTests.cs]: can use [DataRow] here for the 3

File: src/NvGet.Tests/Tools/Updater/PropertyPackageMappingsTests.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+			Assert.IsFalse(result);
+			Assert.IsNull(fallbackPackageId);
+			Assert.IsNull(reason);
+		}
+
+		[TestMethod]
+		public void GivenUnoToolkit_WhenTryGetFallbackMapping_ReturnsFalse()
+		{
+			// UnoToolkit / UnoWinui / UnoResizetizer / UnoTemplates intentionally not mapped:
```
Comment (Xiaoy312) [inline src/NvGet.Tests/Tools/Updater/PropertyPackageMappingsTests.cs]: DataRow

File: src/NvGet/Tools/Updater/Extensions/UpdaterParametersExtension.cs [hunk 6]
```diff
@@ -124,6 +170,34 @@ PackageReference reference
...
+		{
+			if(string.IsNullOrEmpty(packageId))
+			{
+				return packageId;
+			}
+
+			var parts = packageId.Split('.');
+			return string.Concat(parts.Select(p => p.Length == 0
+				? string.Empty
+				: char.ToUpperInvariant(p[0]) + (p.Length > 1 ? p.Substring(1) : string.Empty)));
```
Comment (Xiaoy312) [inline src/NvGet/Tools/Updater/Extensions/UpdaterParametersExtension.cs]:
```suggestion
			var parts = packageId.Split('.', StringSplitOptions.RemoveEmptyEntries);
			return string.Concat(parts.Select(p => char.ToUpperInvariant(p[0]) + p[1..]));
```

File: src/NvGet/Tools/Updater/Entities/PropertyPackageMappings.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		/// Value: tuple of (canonical package ID to look up instead, human-readable reason for logs).
+		/// </summary>
+		/// <remarks>
+		/// Kept private so the table is genuinely immutable from the outside; the only supported access
+		/// path is <see cref="TryGetFallbackMapping(string, out string, out string)"/>. New entries should
+		/// be added here (with a one-line comment + one test method) when a new property→package
+		/// mismatch is observed in production canary build logs.
+		/// </remarks>
+		private static readonly IReadOnlyDictionary<string, (string FallbackPackageId, string Reason)> WellKnownMappings =
+			new Dictionary<string, (string, string)>
```
Comment (Xiaoy312) [inline src/NvGet/Tools/Updater/Entities/PropertyPackageMappings.cs]: your input/key, nuget package id, is not case-sensitive
```suggestion
			new Dictionary<string, (string, string)>(StringComparer.InvariantCultureIgnoreCase)
```

File: src/NvGet/Tools/Updater/Extensions/UpdaterParametersExtension.cs (hunk 6 above)
Comment (Xiaoy312) [reply src/NvGet/Tools/Updater/Extensions/UpdaterParametersExtension.cs]: p[1..] is safe when you dont have string.Empty, which is guaranteed by RemoveEmptyEntries

**Review by Xiaoy312 — APPROVED** [review]

**Review by ajpinedam — APPROVED** [review]
